pkg/agent/cniserver/ipam/antrea_ipam.go

```diff
@@ -195,9 +196,8 @@ func (d *AntreaIPAM) Check(args *invoke.Args, k8sArgs *argtypes.K8sArgs, network
 // of today). If annotation is not present, or annotated IP Pool not found, the driver
 // will not own the request and fall back to next IPAM driver.
 func (d *AntreaIPAM) owns(k8sArgs *argtypes.K8sArgs) (bool, *poolallocator.IPPoolAllocator, []net.IP, *crdv1a2.IPAddressOwner, error) {
-	if d.controller == nil {
-		klog.Warningf("Antrea IPAM driver failed to initialize due to inconsistent configuration. Falling back to default IPAM")
-		return false, nil, nil, nil, nil
+	if err := d.controller.waitControllerReadyForCNIRequest(); err != nil {
```

Comment:
Please add comments to explain why this is needed.

Reply:
How about moving `antreaIPAMDriver.setController(antreaIPAMController)` to controller `Run` function? In this case there will be no informers will be guaranteed to be synced once controller is set in the driver.

Reply:
@annakhm Currently `antreaIPAMDriver.setController(antreaIPAMController)` is called by `InitializeAntreaIPAMController`, and this is OK.
`waitControllerReadyForCNIRequest` will check the controller/informers are not nil and wait a small period (shorter than CNI request timeout=120s)

Reply:
@jianjuns This is a bugfix in https://github.com/antrea-io/antrea/pull/3314 . I added comments to that PR and cherry-picked here.

Reply:
@gran-vmv : could you explain what is the cons. of @annakhm 's idea? In general, I do not like to introduce async wait either. Just hard to track.

Reply:
@jianjuns This is to fix an IP leak for CmdDel called when antrea-agent restarting, which may break the flexible-ipam e2e test.
PR https://github.com/antrea-io/antrea/pull/3314 fixed 2 IP leak issues as mentioned in description, and I cherry-picked these 2 commits here to avoid the e2e fail. You can review the change at https://github.com/antrea-io/antrea/pull/3314 .

Reply:
@jianjuns As I mentioned above, currently `setController` is called by `InitializeAntreaIPAMController`, which is earlier than `Run`, thus we don't need to move it to `Run`.
Since the `CNIServer.Run` is called earlier than `InitializeAntreaIPAMController`, we can only check the controller status when CmdDel called. And a simpler way is to return error immediately instead of wait a small period when controller is not ready.

Reply:
Just to clarify my meaning - async wait would still be needed if `setController` is moved, but we'll only have to wait for controller to be set, without waiting for informers.

Reply:
@annakhm Got your point. Thanks. FIxed.